ec2/utils.py

```diff
@@ -13,61 +13,200 @@
 #    * See the License for the specific language governing permissions and
 #    * limitations under the License.
 
-# built-in imports
+# Built-in Imports
+import os
 import time
 
-# other imports
-from boto.exception import EC2ResponseError, BotoServerError
+# Third-party Imports
+import boto.exception
 
-# Cloudify imports
+# Cloudify Imports
 from ec2 import connection
-from cloudify.exceptions import NonRecoverableError
+from cloudify.exceptions import NonRecoverableError, RecoverableError
+
+
+def get_parameters(ctx):
+    """ These are the list of supported arguments to the run_instances
+        function and their default values. Essentially, this checks to see
+        if the user provided a value and takes the user's value. If the
+        user did not provide a value, the key is deleted from the
+        dictionary.
+      :param ctx:  The Cloudify ctx context. This function uses the
+                   node.properties attribute.
+      :returns parameters dictionary
+    """
 
+    attached_group_ids = get_attached_security_group_ids(ctx=ctx)
+
+    parameters = {
+        'image_id': None, 'key_name': None, 'security_groups': None,
+        'user_data': None, 'addressing_type': None,
+        'instance_type': 'm1.small', 'placement': None, 'kernel_id': None,
+        'ramdisk_id': None, 'monitoring_enabled': False, 'subnet_id': None,
+        'block_device_map': None, 'disable_api_termination': False,
+        'instance_initiated_shutdown_behavior': None,
+        'private_ip_address': None, 'placement_group': None,
+        'client_token': None, 'security_group_ids': None,
+        'additional_info': None, 'instance_profile_name': None,
+        'instance_profile_arn': None, 'tenancy': None, 'ebs_optimized': False,
+        'network_interfaces': None, 'dry_run': False
+    }
+
+    node_parameter_keys = ctx.node.properties['parameters'].keys()
+
+    for key in parameters.keys():
+        if key is 'security_group_ids':
+            if key in node_parameter_keys:
+                parameters[key] = list(
+                    set(attached_group_ids) | set(
+                        ctx.node.properties['parameters'][key])
+                )
+            else:
+                parameters[key] = attached_group_ids
+        elif key is 'key_name':
+            if key in node_parameter_keys:
+                parameters[key] = ctx.node.properties['parameters'][key]
+            else:
+                parameters[key] = get_attached_keypair_id(ctx)
+        elif key in node_parameter_keys:
+            parameters[key] = ctx.node.properties['parameters'][key]
+        elif key is 'image_id' or key is 'instance_type':
+            parameters[key] = ctx.node.properties[key]
+        else:
+            del(parameters[key])
+
+    return parameters
+
+
+def validate_node_property(key, ctx):
```

Comment:
Why do we need to validate that a property exists within ctx? We can just try and use it from within the task.